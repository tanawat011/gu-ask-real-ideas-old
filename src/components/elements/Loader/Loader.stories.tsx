import { ComponentMeta, ComponentStory } from '@storybook/react'

import { LOADER_STYLE } from '@constants'
import { Loader } from '@element/Loader'

const ComponentAllLoaders: React.FC = () => (
  <div className='flex flex-wrap'>
    <div className='flex flex-col justify-center items-center h-40 w-40 m-4'>
      <Loader />
      Default
    </div>
    <div className='flex flex-col justify-center items-center h-40 w-40 m-4'>
      <Loader loaderStyle={LOADER_STYLE.TWO_PIPE} />2 Pipe
    </div>
    <div className='flex flex-col justify-center items-center h-40 w-40 m-4'>
      <Loader loaderStyle={LOADER_STYLE.THREE_FLOOR} />3 Floor
    </div>
    <div className='flex flex-col justify-center items-center h-40 w-40 m-4'>
      <Loader loaderStyle={LOADER_STYLE.DASH} />
      Dash
    </div>
  </div>
)

export default {
  title: 'Loader',
  component: ComponentAllLoaders,
} as ComponentMeta<typeof ComponentAllLoaders>

const Template: ComponentStory<typeof Loader> = (args) => (
  <ComponentAllLoaders {...args} />
)

export const Default = Template.bind({})
