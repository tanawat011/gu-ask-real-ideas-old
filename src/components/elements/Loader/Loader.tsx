import clsx from 'clsx'

import { LOADER_COLOR, LOADER_STYLE } from '@constants'

type PropsLoader = {
  color?: LOADER_COLOR
  loaderStyle?: LOADER_STYLE
  className?: string
}

const Loader3FloorStyle = (classColor: string) => (
  <svg
    className={clsx('loader-3floor', classColor)}
    version='1.1'
    id='L7'
    x='0px'
    y='0px'
    viewBox='0 0 100 100'
    enableBackground='new 0 0 100 100'
  >
    <path
      d='M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3
c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z'
    >
      <animateTransform
        attributeName='transform'
        attributeType='XML'
        type='rotate'
        dur='2s'
        from='0 50 50'
        to='360 50 50'
        repeatCount='indefinite'
      />
    </path>
    <path
      d='M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7
c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z'
    >
      <animateTransform
        attributeName='transform'
        attributeType='XML'
        type='rotate'
        dur='1s'
        from='0 50 50'
        to='-360 50 50'
        repeatCount='indefinite'
      />
    </path>
    <path
      d='M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5
L82,35.7z'
    >
      <animateTransform
        attributeName='transform'
        attributeType='XML'
        type='rotate'
        dur='2s'
        from='0 50 50'
        to='360 50 50'
        repeatCount='indefinite'
      />
    </path>
  </svg>
)

const LoaderDashStyle = (classColor: string) => (
  <svg className={clsx('loader-dash', classColor)} viewBox='0 0 50 50'>
    <circle
      className='path'
      cx='25'
      cy='25'
      r='20'
      fill='none'
      strokeWidth='8'
    ></circle>
  </svg>
)

export const Loader: React.FC<PropsLoader> = (props) => {
  const {
    loaderStyle = LOADER_STYLE.DEFAULT,
    color = LOADER_COLOR.DEFAULT,
    className = '',
  } = props

  const renderColor = (style: LOADER_STYLE, classColor: LOADER_COLOR) => {
    if (style === LOADER_STYLE.TWO_PIPE) {
      switch (classColor) {
        case LOADER_COLOR.SKY:
          return 'border-t-nextui-blue-400 border-b-nextui-blue-400'
        case LOADER_COLOR.WHITE:
          return 'border-t-white border-b-white'

        default:
          return 'border-t-nextui-gray-400 border-b-nextui-gray-400'
      }
    }

    if (style === LOADER_STYLE.THREE_FLOOR) {
      switch (classColor) {
        case LOADER_COLOR.SKY:
          return 'fill-nextui-blue-400'
        case LOADER_COLOR.WHITE:
          return 'fill-white'

        default:
          return 'fill-nextui-gray-400'
      }
    }

    if (style === LOADER_STYLE.DASH) {
      switch (classColor) {
        case LOADER_COLOR.SKY:
          return 'stroke-nextui-blue-400'
        case LOADER_COLOR.WHITE:
          return 'stroke-white'

        default:
          return 'stroke-nextui-gray-400'
      }
    }

    switch (classColor) {
      case LOADER_COLOR.SKY:
        return 'border-t-nextui-blue-400'
      case LOADER_COLOR.WHITE:
        return 'border-t-white'

      default:
        return 'border-t-nextui-gray-400'
    }
  }

  const renderLoader = (style: LOADER_STYLE, classColor: string) => {
    switch (style) {
      case LOADER_STYLE.TWO_PIPE:
        return (
          <div
            className={clsx(
              'loader-second border-transparent border-4 border-t-4 border-b-4',
              renderColor(LOADER_STYLE.DEFAULT, color),
              classColor,
            )}
          />
        )

      case LOADER_STYLE.DASH:
        return LoaderDashStyle(classColor)

      case LOADER_STYLE.THREE_FLOOR:
        return Loader3FloorStyle(classColor)

      case LOADER_STYLE.DEFAULT:
      default:
        return (
          <div
            className={clsx(
              'loader border-transparent border-4 border-t-4',
              classColor,
            )}
          />
        )
    }
  }

  const classColor = renderColor(loaderStyle, color)
  const loader = renderLoader(loaderStyle, classColor)

  return <div className={className}>{loader}</div>
}
