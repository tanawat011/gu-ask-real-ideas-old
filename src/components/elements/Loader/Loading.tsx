import { LOADING_COLOR, LOADING_STYLE } from '@constants'
import { Loading3FloorStyle, LoadingDashStyle } from '@element/svg'

interface PropsLoading {
  color?: LOADING_COLOR
  loaderStyle?: LOADING_STYLE
  className?: string
}

export const Loading: React.FC<PropsLoading> = (props) => {
  const {
    loaderStyle = LOADING_STYLE.DEFAULT,
    color = LOADING_COLOR.DEFAULT,
    className = '',
  } = props

  const renderColor = (classColor: LOADING_COLOR) => {
    if (loaderStyle === LOADING_STYLE.TWO_PIPE) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'border-t-nextui-blue-400 border-b-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'border-t-white border-b-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'border-t-nextui-gray-400 border-b-nextui-gray-400'
      }
    }

    if (loaderStyle === LOADING_STYLE.THREE_FLOOR) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'fill-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'fill-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'fill-nextui-gray-400'
      }
    }

    if (loaderStyle === LOADING_STYLE.DASH) {
      switch (classColor) {
        case LOADING_COLOR.SKY:
          return 'stroke-nextui-blue-400'
        case LOADING_COLOR.WHITE:
          return 'stroke-white'

        case LOADING_COLOR.DEFAULT:
        default:
          return 'stroke-nextui-gray-400'
      }
    }

    switch (classColor) {
      case LOADING_COLOR.SKY:
        return 'border-t-nextui-blue-400'
      case LOADING_COLOR.WHITE:
        return 'border-t-white'

      case LOADING_COLOR.DEFAULT:
      default:
        return 'border-t-nextui-gray-400'
    }
  }

  const renderLoader = (classColor: string) => {
    const baseClass = 'border-transparent border-4 border-t-4'

    switch (loaderStyle) {
      case LOADING_STYLE.TWO_PIPE:
        return (
          <div
            className={`loader-second ${baseClass} border-b-4 ${renderColor(
              color,
            )} ${classColor}`}
          />
        )

      case LOADING_STYLE.DASH:
        return LoadingDashStyle(classColor)

      case LOADING_STYLE.THREE_FLOOR:
        return Loading3FloorStyle(classColor)

      case LOADING_STYLE.DEFAULT:
      default:
        return <div className={`loader ${baseClass} ${classColor}`} />
    }
  }

  const classColor = renderColor(color)
  const loader = renderLoader(classColor)

  return <div className={className}>{loader}</div>
}
