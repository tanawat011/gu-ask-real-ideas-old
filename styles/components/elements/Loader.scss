/**
 * NOTE: Default loader and second style loader
 */
.loader, %loader {
  width: 24px;
  height: 24px;
  border: 4px solid #f3f3f3; /* Background */
  border-radius: 50%; /* Rounded borders */
  border-top: 4px solid #3498db; /* Loader */
  -webkit-animation: spin 1s linear infinite; /* Safari */
  animation: spin 1s linear infinite; /* spin | speed | spin style | time for spin */
  z-index: 2;

  &-second {  
    @extend %loader;

    border-bottom: 4px solid #3498db;
  }
}

/* Safari */
@-webkit-keyframes spin {
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

/**
 * NOTE: Dash style loader
 */
.loader-dash {
  width: 24px;
  height: 24px;
  animation: rotate 2s linear infinite;
  z-index: 2;
  
  & .path {
    stroke: #3498db;
    stroke-linecap: round;
    animation: dash 1.5s ease-in-out infinite;
  }
}

@keyframes rotate {
  100% {
    transform: rotate(360deg);
  }
}

@keyframes dash {
  0% {
    stroke-dasharray: 1, 150;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -35;
  }

  100% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -124;
  }
}

/**
 * 3 Floor style loader
 */
.loader-3floor {
  width: 24px;
  height: 24px;
  fill: #3498db;
}
