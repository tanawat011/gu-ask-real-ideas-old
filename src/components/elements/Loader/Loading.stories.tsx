import { ComponentMeta, ComponentStory } from '@storybook/react'
import clsx from 'clsx'

import { LOADING_COLOR, LOADING_STYLE } from '@constants'
import { Loading } from '@element/Loader'

const ComponentAllLoadings: React.FC<React.ComponentProps<typeof Loading>> = (
  props,
) => {
  const classCard = clsx(
    'flex',
    'flex-col',
    'justify-center',
    'items-center',
    'h-28',
    'w-28',
    'm-4',
    'bg-nextui-gray-700',
    'rounded-md',
    'leading-6',
  )

  return (
    <div className='flex flex-wrap'>
      <div className={classCard}>
        <Loading
          className='mb-5'
          {...props}
          loaderStyle={LOADING_STYLE.DEFAULT}
        />

        <span className='text-white'>Default</span>
      </div>

      <div className={classCard}>
        <Loading
          className='mb-5'
          {...props}
          loaderStyle={LOADING_STYLE.TWO_PIPE}
        />

        <span className='text-white'>2 Pipe</span>
      </div>

      <div className={classCard}>
        <Loading
          className='mb-5'
          {...props}
          loaderStyle={LOADING_STYLE.THREE_FLOOR}
        />

        <span className='text-white'>3 Floor</span>
      </div>

      <div className={classCard}>
        <Loading className='mb-5' {...props} loaderStyle={LOADING_STYLE.DASH} />

        <span className='text-white'>Dash</span>
      </div>
    </div>
  )
}

export default {
  title: 'Loading',
  component: Loading,
} as ComponentMeta<typeof Loading>

const Template: ComponentStory<typeof Loading> = (args) => <Loading {...args} />

export const Default = Template.bind({})

Default.args = {
  color: LOADING_COLOR.DEFAULT,
}

const TemplateAll: ComponentStory<typeof Loading> = () => (
  <>
    <ComponentAllLoadings color={LOADING_COLOR.DEFAULT} />
    <ComponentAllLoadings color={LOADING_COLOR.SKY} />
    <ComponentAllLoadings color={LOADING_COLOR.WHITE} />
  </>
)

export const All = TemplateAll.bind({})
