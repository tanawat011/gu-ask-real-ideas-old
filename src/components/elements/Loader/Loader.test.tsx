import { render } from '@testing-library/react'
import { Loader } from '@element/Loader'
import { LOADER_STYLE } from '@constants'

describe('<Loader />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render to match snapshot with default style', () => {
    expect.assertions(1)

    const { asFragment } = render(<Loader />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render to match snapshot with default style 2', () => {
    expect.assertions(1)

    const { asFragment } = render(<Loader loaderStyle={LOADER_STYLE.DEFAULT} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render to match snapshot with second style', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <Loader loaderStyle={LOADER_STYLE.TWO_PIPE} />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render to match snapshot with dash style', () => {
    expect.assertions(1)

    const { asFragment } = render(<Loader loaderStyle={LOADER_STYLE.DASH} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render to match snapshot with 3floor style', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <Loader loaderStyle={LOADER_STYLE.THREE_FLOOR} />,
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
