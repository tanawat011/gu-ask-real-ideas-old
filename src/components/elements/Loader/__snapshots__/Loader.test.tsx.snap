// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> should render to match snapshot with 3floor style 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="loader-3floor fill-gray-500"
      enable-background="new 0 0 100 100"
      id="L7"
      version="1.1"
      viewBox="0 0 100 100"
      x="0px"
      y="0px"
    >
      <path
        d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="1s"
          from="0 50 50"
          repeatCount="indefinite"
          to="-360 50 50"
          type="rotate"
        />
      </path>
      <path
        d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="2s"
          from="0 50 50"
          repeatCount="indefinite"
          to="360 50 50"
          type="rotate"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loader /> should render to match snapshot with dash style 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="loader-dash stroke-gray-500"
      viewBox="0 0 50 50"
    >
      <circle
        class="path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke-width="8"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Loader /> should render to match snapshot with default style 1`] = `
<DocumentFragment>
  <div>
    <div
      class="loader border-4 border-t-4 border-t-gray-500"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loader /> should render to match snapshot with default style 2 1`] = `
<DocumentFragment>
  <div>
    <div
      class="loader border-4 border-t-4 border-t-gray-500"
    />
  </div>
</DocumentFragment>
`;

exports[`<Loader /> should render to match snapshot with second style 1`] = `
<DocumentFragment>
  <div>
    <div
      class="loader-second border-4 border-t-4 border-b-4 border-t-gray-500 border-b-gray-500"
    />
  </div>
</DocumentFragment>
`;
